File: automation-test-selenium/src/main/java/org/suntower/pages/admin/AdminDashboardPage.java
package org.suntower.pages.admin;

import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.suntower.pages.core.AdminShellPage;

public class AdminDashboardPage extends AdminShellPage {
  public AdminDashboardPage(WebDriver driver) {
    super(driver);
    this.path = "/admin/dashboard";
  }

  public void waitForLoaded() {
    waitForUrlContains("/admin/dashboard");
    waitForVisible(css("#totalBuildingsStat"));
  }

  public void waitForOverviewVisible() {
    waitForVisible(css("#totalBuildingsStat"));
    waitForVisible(css("#totalCustomersStat"));
    waitForVisible(css("#totalStaffsStat"));
    waitForVisible(css("#totalContractsStat"));
    waitForVisible(css("#revenueChartCombined"));
    waitForVisible(css("#contractsByBuildingChart"));
    waitForVisible(css("#districtChart"));
    waitForVisible(css("#saleRateChart"));
    waitForVisible(css("#potentialCustomersBody"));
    waitForVisible(css("#topStaffsBody"));
    waitForVisible(css("#recentBuildingsContainer"));
  }

  public void openBuildingsFromStatCard() {
    clickStatCard("totalBuildingsStat");
    waitForUrlContains("/admin/building/list");
  }

  public void openCustomersFromStatCard() {
    clickStatCard("totalCustomersStat");
    waitForUrlContains("/admin/customer/list");
  }

  public void openStaffsFromStatCard() {
    clickStatCard("totalStaffsStat");
    waitForUrlContains("/admin/staff/list");
  }

  public void openContractsFromStatCard() {
    clickStatCard("totalContractsStat");
    waitForUrlContains("/admin/contract/list");
  }

  public void waitForRecentBuildingVisible(String buildingName) {
    waitForText(css("#recentBuildingsContainer"), buildingName);
  }

  public void openRecentBuilding(String buildingName) {
    By item =
        By.xpath(
            "//*[@id='recentBuildingsContainer']//*[contains(@class,'recent-item') and contains(.,'"
                + buildingName
                + "')]");
    click(item);
  }

  private By statCardByStatId(String statId) {
    return By.xpath("//*[@id='" + statId + "']/ancestor::div[contains(@class,'stat-card')]");
  }

  private void clickStatCard(String statId) {
    WebElement statCard = visible(statCardByStatId(statId));
    scrollIntoView(statCard);
    ((JavascriptExecutor) driver).executeScript("arguments[0].click();", statCard);
  }
}

File: automation-test-playwright/pages/admin/AdminDashboardPage.ts
import { type Locator } from "@playwright/test";
import { AdminShellPage } from "../core/AdminShellPage";

export class AdminDashboardPage extends AdminShellPage {
  protected readonly path = "/admin/dashboard";

  private statCardByStatId(statId: string): Locator {
    return this.page.locator(`#${statId}`).locator("xpath=ancestor::div[contains(@class,'stat-card')]");
  }

  async waitForLoaded(): Promise<void> {
    await this.page.waitForURL(/\/admin\/dashboard/);
    await this.waitForVisible(this.page.locator("#totalBuildingsStat"));
  }

  async waitForOverviewVisible(): Promise<void> {
    await this.waitForVisible(this.page.locator("#totalBuildingsStat"));
    await this.waitForVisible(this.page.locator("#totalCustomersStat"));
    await this.waitForVisible(this.page.locator("#totalStaffsStat"));
    await this.waitForVisible(this.page.locator("#totalContractsStat"));
    await this.waitForVisible(this.page.locator("#revenueChartCombined"));
    await this.waitForVisible(this.page.locator("#contractsByBuildingChart"));
    await this.waitForVisible(this.page.locator("#districtChart"));
    await this.waitForVisible(this.page.locator("#saleRateChart"));
    await this.waitForVisible(this.page.locator("#potentialCustomersBody"));
    await this.waitForVisible(this.page.locator("#topStaffsBody"));
    await this.waitForVisible(this.page.locator("#recentBuildingsContainer"));
  }

  async openBuildingsFromStatCard(): Promise<void> {
    await this.statCardByStatId("totalBuildingsStat").click();
  }

  async openCustomersFromStatCard(): Promise<void> {
    await this.statCardByStatId("totalCustomersStat").click();
  }

  async openStaffsFromStatCard(): Promise<void> {
    await this.statCardByStatId("totalStaffsStat").click();
  }

  async openContractsFromStatCard(): Promise<void> {
    await this.statCardByStatId("totalContractsStat").click();
  }

  async waitForRecentBuildingVisible(buildingName: string): Promise<void> {
    await this.waitForVisible(
      this.firstVisible(this.page.locator("#recentBuildingsContainer .recent-item").filter({ hasText: buildingName }))
    );
  }

  async openRecentBuilding(buildingName: string): Promise<void> {
    await this.firstVisible(
      this.page.locator("#recentBuildingsContainer .recent-item").filter({ hasText: buildingName })
    ).click();
  }
}
